lib/mayaUsd/ufe/UsdUndoCreateStageWithNewLayerCommand.cpp

```diff
@@ -0,0 +1,211 @@
+//
+// Copyright 2023 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdUndoCreateStageWithNewLayerCommand.h"
+
+#include <mayaUsd/ufe/UsdStageMap.h>
+#include <mayaUsd/ufe/UsdUndoRenameCommand.h>
+#include <mayaUsd/ufe/Utils.h>
+#include <mayaUsd/undo/OpUndoItems.h>
+
+#include <ufe/hierarchy.h>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+extern UsdStageMap g_StageMap;
+
+UsdUndoCreateStageWithNewLayerCommand::UsdUndoCreateStageWithNewLayerCommand(
+    const Ufe::SceneItem::Ptr& parentItem)
+    : _parentItem(nullptr)
+    , _insertedChild(nullptr)
+    , _createTransformDagMod(MDagModifierUndoItem::create("Create transform"))
+    , _createProxyShapeDagMod(MDagModifierUndoItem::create("Create Stage with new Layer"))
+    , _createTransformSuccess(false)
+    , _createProxyShapeSuccess(false)
```

Comment:
There should be just a single success member variable. The entire execute() method should be atomic so that at any point if there is a failure you must undo everything you've done up to that point. For instance if you failure creating the mayaUsdProxyShape node, then you must delete the transform. Then both the undo/redo simply have an early out if the success variable is not true.

Reply:
True, I'm missing something like the `markAsFailed()` methods of the other commands. Good catch!

However, I think I still need two booleans because there are two MDagModifiers. 

The documentation of `MDagModifier::undoIt()` states:
> It is only valid to call this method after the doIt method has been called.

If the `execute()` method fails, I have to undo all the `doIt()` calls that have already been executed but I must not call `undoIt()` on DAG modifiers whose `doIt()` method has not been executed yet. Thus, I have to remember for every `MDagModifier` if its `doIt()` method has been called.

I added a `markAsFailed()` method which uses both booleans and I made `undo()` and `redo()` rely on a single boolean. Let me know if that's okay. See ab589dc3bb7610cabf1b87e8e19d54050edb7328.

Reply:
The idea is that you should call `undoIt()` immediately in the `execute()` method, so that it is atomic: either it completely succeeds, or any of its effects are undone.  That way, you only need a single boolean to indicate success of the `execute()` method, and you can early out of `undo()` and `redo()`

Reply:
I see. Like this? fd207b80b1a4492a96269d99bd704233030e9277

Reply:
Made use of try/catch for the undo cleanup. See 8c3a19616a537596377ee4c7b4d4e0aff25108eb.